In [1]:
import torch

# BCELoss
在图片多标签分类时，如果3张图片分三类，那么会输出一个3*3的矩阵

In [13]:
input = torch.FloatTensor([[-0.4089, -1.2471, 0.5907],
                          [-0.4897, -0.8267, -0.7349],
                          [0.5241, -0.1246, -0.4751]])
input

tensor([[-0.4089, -1.2471,  0.5907],
        [-0.4897, -0.8267, -0.7349],
        [ 0.5241, -0.1246, -0.4751]])

In [14]:
m = torch.nn.Sigmoid()
input = m(input)
input

tensor([[0.3992, 0.2232, 0.6435],
        [0.3800, 0.3043, 0.3241],
        [0.6281, 0.4689, 0.3834]])

假如Target是：

In [15]:
target = torch.FloatTensor([[0,1,1],
                           [0,0,1],
                           [1,0,1]])


BCELoss是 $-\frac{1}{n}\sum (y_n \times ln x_n) + (1-y_n) \times ln(1-x_n)$


其中,y是target，x是模型输出的值


所以对于第一行：

第一列0×ln0.3992+(1−0)×ln(1−0.3992)=−0.50950×ln0.3992+(1−0)×ln(1−0.3992)=−0.5095

第二列1×ln0.2232+(1−1)×ln(1−0.2232)=−1.49971×ln0.2232+(1−1)×ln(1−0.2232)=−1.4997

第三列1×ln0.6435+(1−1)×ln(1−0.6435)=−0.44081×ln0.6435+(1−1)×ln(1−0.6435)=−0.4408


第二行：

第一列0×ln0.3800+(1−0)×ln(1−0.3800)=−0.47800×ln0.3800+(1−0)×ln(1−0.3800)=−0.4780

第二列0×ln0.3044+(1−0)×ln(1−0.3044)=−0.36300×ln0.3044+(1−0)×ln(1−0.3044)=−0.3630

第三列1×ln0.3241+(1−1)×ln(1−0.3241)=−1.12671×ln0.3241+(1−1)×ln(1−0.3241)=−1.1267


第三行：

第一列1×ln0.6281+(1−1)×ln(1−0.6281)=−0.46511×ln0.6281+(1−1)×ln(1−0.6281)=−0.4651

第二列0×ln0.4689+(1−0)×ln(1−0.4689)=−0.63280×ln0.4689+(1−0)×ln(1−0.4689)=−0.6328

第三列1×ln0.3834+(1−1)×ln(1−0.3834)=−0.95871×ln0.3834+(1−1)×ln(1−0.3834)=−0.9587


去掉负号求均值：

$\frac{0.5095+1.4997+0.4408}{3} = 0.8167$

$\frac{0.4780+0.3630+1.1267}{3} = 0.6559$

$\frac{0.4651+0.6328+0.9587}{3} = 0.6855$ 

在求个平均：

$\frac{0.8167+0.6559+0.6855}{3} = 0.7194$

下面我们用BCELoss来验证一下Loss是不是0.7194

In [16]:
loss = torch.nn.BCELoss()

loss(input, target)

tensor(0.7193)

这里的误差0.0001是因为上面计算损失值时，每次都只保留了四位小数，所以存在了一点误差。

# BCEWithLogitsLoss

**BCEWithLogitsLoss就是把sigmoid-BCELoss合成了一步完成，我们直接用刚刚的input验证一下结果是不是0.7193**

In [18]:
input = torch.FloatTensor([[-0.4089, -1.2471, 0.5907],
                          [-0.4897, -0.8267, -0.7349],
                          [0.5241, -0.1246, -0.4751]])

target = torch.FloatTensor([[0,1,1],
                           [0,0,1],
                           [1,0,1]])

loss = torch.nn.BCEWithLogitsLoss()
loss(input, target)



tensor(0.7193)

## 其他的例子

In [28]:
x = torch.rand([2,3,4])
labels = torch.round(x)
# torch.round()返回相邻的最大整数
labels

tensor([[[0., 1., 1., 0.],
         [1., 1., 1., 1.],
         [0., 0., 1., 1.]],

        [[1., 0., 1., 0.],
         [0., 1., 0., 0.],
         [0., 1., 0., 1.]]])

In [31]:
input = torch.randn([2,3,4])
input

tensor([[[-0.0828,  0.1661, -0.5422, -1.2300],
         [-1.0764, -0.1699, -0.0334,  0.3021],
         [-0.3994, -0.8070,  1.0112, -0.7639]],

        [[-1.4431,  0.3587,  0.2310, -0.7968],
         [ 1.0035,  1.2438, -0.0235, -1.0956],
         [-1.6103,  2.2160,  0.5243,  0.4291]]])

In [33]:
loss_layer = torch.nn.BCEWithLogitsLoss()
loss = loss_layer(input, labels)
loss

tensor(0.6705)